File: fe2-android/app/src/main/java/com/github/dedis/student20_pop/model/network/method/message/data/election/ElectionSetup.java
package com.github.dedis.student20_pop.model.network.method.message.data.election;

import com.github.dedis.student20_pop.model.network.method.message.data.Action;
import com.github.dedis.student20_pop.model.network.method.message.data.Data;
import com.github.dedis.student20_pop.model.network.method.message.data.Objects;
import com.github.dedis.student20_pop.utility.network.IdGenerator;
import com.google.gson.annotations.SerializedName;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

public class ElectionSetup extends Data {

    private final String id;
    private final String name;
    private final String lao;
    @SerializedName(value = "created_at")
    private final long createdAt;
    @SerializedName(value = "start_time")
    private final long startTime;
    @SerializedName(value = "end_time")
    private final long endTime;
    private final String version;
    private final List<ElectionQuestion> questions;

    /**
     * Constructor for a data setup Election Event
     *
     * @param name  name of the Election
     * @param start of the Election
     * @param laoId id of the LAO
     */
    public ElectionSetup(
            String name,
            long start,
            long end,
            String votingMethod,
            boolean writeIn,
            List<String> ballotOptions,
            String question,
            String laoId) {
        this.name = name;
        this.createdAt = Instant.now().getEpochSecond();
        this.startTime = start;
        this.endTime = end;
        this.lao = laoId;
        this.version = "1.0.0";
        this.id = IdGenerator.generateElectionSetupId(laoId, createdAt, name);
        this.questions = new ArrayList<>();
        this.questions.add(new ElectionQuestion(question, votingMethod, writeIn, ballotOptions, this.id));
    }


    public String getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public long getCreation() {
        return createdAt;
    }

    public long getStartTime() {
        return startTime;
    }

    public long getEndTime() {
        return endTime;
    }

    public List<ElectionQuestion> getQuestions() {
        return questions;
    }

    public String getLao() {
        return lao;
    }

    public String getVersion() {
        return version;
    }


    @Override
    public String getObject() {
        return Objects.ELECTION.getObject();
    }

    @Override
    public String getAction() {
        return Action.SETUP.getAction();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        ElectionSetup that = (ElectionSetup) o;
        return getCreation() == that.getCreation()
                && startTime == that.getStartTime()
                && java.util.Objects.equals(getId(), that.getId())
                && createdAt == that.getCreation()
                && java.util.Objects.equals(getName(), that.getName())
                && endTime == that.getEndTime()
                && java.util.Objects.equals(questions, that.getQuestions());
    }

    @Override
    public int hashCode() {
        return java.util.Objects.hash(
                getId(),
                getName(),
                getCreation(),
                getStartTime(),
                getEndTime(),
                getQuestions());
    }

    @Override
    public String toString() {
        return "ElectionSetup{"
                + "id='"
                + id
                + '\''
                + ", name='"
                + name
                + '\''
                + ", lao='"
                + lao
                + '\''
                + ", creation='"
                + createdAt
                + '\''
                + ", start='"
                + startTime
                + '\''
                + ", end="
                + endTime
                + '\''
                + ", version='"
                + version
                + '\''
                + questions.get(0).toString()
                + '}';
    }

}

File: be1-go/internal/handler/channel/lao/mlao/election_setup.go
package mlao

import (
	"encoding/base64"
	"popstellar/internal/errors"
	"popstellar/internal/handler/channel"
	"strconv"
)

// ElectionSetup defines a message data
type ElectionSetup struct {
	Object  string `json:"object"`
	Action  string `json:"action"`
	ID      string `json:"id"`
	Lao     string `json:"lao"`
	Name    string `json:"name"`
	Version string `json:"version"`

	// CreatedAt is a Unix timestamp
	CreatedAt int64 `json:"created_at"`

	// StartTime is a Unix timestamp
	StartTime int64 `json:"start_time"`

	// EndTime is a Unix timestamp
	EndTime int64 `json:"end_time"`

	Questions []ElectionSetupQuestion `json:"questions"`
}

const ElectionFlag = "Election"

func (message ElectionSetup) Verify(laoID string) error {
	_, err := base64.URLEncoding.DecodeString(message.Lao)
	if err != nil {
		return errors.NewInvalidMessageFieldError("failed to decode lao: %v", err)
	}

	if message.Lao != laoID {
		return errors.NewInvalidMessageFieldError("lao id is %s, should be %s", message.Lao, laoID)
	}

	_, err = base64.URLEncoding.DecodeString(message.ID)
	if err != nil {
		return errors.NewInvalidMessageFieldError("failed to decode election id: %v", err)
	}

	// verify election setup message id
	expectedID := channel.Hash(
		ElectionFlag,
		laoID,
		strconv.Itoa(int(message.CreatedAt)),
		message.Name,
	)
	if message.ID != expectedID {
		return errors.NewInvalidMessageFieldError("election id is %s, should be %s", message.ID, expectedID)
	}
	if len(message.Name) == 0 {
		return errors.NewInvalidMessageFieldError("election name is empty")
	}
	if message.Version != OpenBallot && message.Version != SecretBallot {
		return errors.NewInvalidMessageFieldError("election version is %s, should be %s or %s", message.Version,
			OpenBallot, SecretBallot)
	}
	if message.CreatedAt < 0 {
		return errors.NewInvalidMessageFieldError("election created at is %d, should be minimum 0", message.CreatedAt)
	}
	if message.StartTime < message.CreatedAt {
		return errors.NewInvalidMessageFieldError("election start should be greater that creation time")
	}
	if message.EndTime < message.StartTime {
		return errors.NewInvalidMessageFieldError("election end should be greater that start time")
	}
	if len(message.Questions) == 0 {
		return errors.NewInvalidMessageFieldError("election contains no questions")
	}
	return nil
}

// GetObject implements MessageData
func (ElectionSetup) GetObject() string {
	return channel.ElectionObject
}

// GetAction implements MessageData
func (ElectionSetup) GetAction() string {
	return channel.ElectionActionSetup
}

// NewEmpty implements MessageData
func (ElectionSetup) NewEmpty() channel.MessageData {
	return &ElectionSetup{}
}

const (
	// OpenBallot is a type of election
	OpenBallot = "OPEN_BALLOT"
	// SecretBallot is a type of election
	SecretBallot    = "SECRET_BALLOT"
	questionFlag    = "Question"
	PluralityMethod = "Plurality"
	ApprovalMethod  = "Approval"
)

// ElectionSetupQuestion defines a question of an election setup
type ElectionSetupQuestion struct {
	ID            string   `json:"id"`
	Question      string   `json:"question"`
	VotingMethod  string   `json:"voting_method"`
	BallotOptions []string `json:"ballot_options"`
	WriteIn       bool     `json:"write_in"`
}

func (q ElectionSetupQuestion) Verify(electionSetupID string) error {
	_, err := base64.URLEncoding.DecodeString(q.ID)
	if err != nil {
		return errors.NewInvalidMessageFieldError("failed to decode Question id: %v", err)
	}

	expectedID := channel.Hash(
		questionFlag,
		electionSetupID,
		q.Question,
	)

	if q.ID != expectedID {
		return errors.NewInvalidMessageFieldError("Question id is %s, should be %s", q.ID, expectedID)
	}

	if len(q.Question) == 0 {
		return errors.NewInvalidMessageFieldError("Question is empty")
	}

	if q.VotingMethod != PluralityMethod && q.VotingMethod != ApprovalMethod {
		return errors.NewInvalidMessageFieldError("Question voting method is %s, should be %s or %s",
			q.VotingMethod, PluralityMethod, ApprovalMethod)
	}

	return nil
}
